pkg/execution/state/redis_state/redis_state.go

```diff
@@ -0,0 +1,349 @@
+type Opt func(r *mgr)
+
+// New returns a state manager which uses Redis as the backing state store.
+//
+// TODO: Explain approach to storing state.
+func New(opts ...Opt) state.Manager {
+	m := &mgr{
+		kf: defaultKeyFunc{},
+	}
+
+	for _, opt := range opts {
+		opt(m)
+	}
+
+	if m.r == nil {
+		m.r = redis.NewClient(&redis.Options{
+			Addr:     "localhost:6379",
+			Password: "",
+			DB:       0,
+		})
+	}
+
+	return m
+}
+
+func WithConnectOpts(o redis.Options) Opt {
+	return func(m *mgr) {
+		m.r = redis.NewClient(&o)
+	}
+}
+
+// WithKeyGenerator specifies the function to use when creating keys for
+// each stored data type.
+func WithKeyGenerator(kf KeyGenerator) Opt {
+	return func(m *mgr) {
+		m.kf = kf
+	}
+}
+
+// KeyFunc returns a unique string based off of given data, which is used
+// as the key for data stored in redis for workflows, events, actions, and
+// errors.
+type KeyGenerator interface {
+	// TODO: Deterministically generate a v5 UUID based off of the name
+	// of the workflow in dev.
+	Workflow(ctx context.Context, workflowID uuid.UUID, version int) string
+
+	// RunMetadata stores state regarding the current run identifier, such
+	// as the workflow version, the time the run started, etc.
+	RunMetadata(context.Context, state.Identifier) string
+
+	// Event returns the key used to store the specific event for the
+	// given workflow run.
+	Event(context.Context, state.Identifier) string
+
+	// Actions returns the key used to store the action response map used
+	// for given workflow run - ie. the results for individual steps.
+	Actions(context.Context, state.Identifier) string
+
+	// Errors returns the key used to store the error hash map used
+	// for given workflow run.
+	Errors(context.Context, state.Identifier) string
+
+	// Pause returns the key used to store an individual pause.
+	Pause(context.Context, state.Pause) string
+}
+
+type mgr struct {
+	kf KeyGenerator
+	r  *redis.Client
+}
+
+func (m mgr) New(ctx context.Context, workflow inngest.Workflow, runID ulid.ULID, input map[string]any) (state.State, error) {
+	id := state.Identifier{
+		WorkflowID: workflow.UUID,
+		RunID:      runID,
+	}
+
+	// TODO: We could probably optimize the commands here by storing the event
+	// within run metadata.  We want step output (actions) and errors to be
+	// their own redis hash for fast inserts (HSET on individual step results).
+	// However, the input event is immutable.
+	metadata := runMetadata{
+		Version:   workflow.Version,
+		CreatedAt: time.Now(),
+	}
+
+	// We marshal this ahead of creating a redis transaction as it's necessary
+	// every time and reduces the duration that the lock is held.
+	inputJSON, err := json.Marshal(input)
+	if err != nil {
+		return nil, err
+	}
+
+	err = m.r.Watch(ctx, func(tx *redis.Tx) error {
+
+		// Ensure that the workflow exists within the state store.
+		key := m.kf.Workflow(ctx, workflow.UUID, workflow.Version)
+		val, err := tx.Exists(ctx, key).Uint64()
```

Comment:
Could this use [`SETNX`](https://redis.io/commands/setnx/) to combine these steps or is it more performant not to have to marshal the workflow every new run?